### PR TITLE
compute mean of a set of rotations

### PR DESCRIPTION
**Pull request by petercorke**
New method `mean()` for each of `SO3` and `UnitQuaternion`, computes the Karcher (rotational) mean.

Relatively straightforward for `SO3` but `UnitQuaternion` required a bunch of changes to allow more methods to handle multi-valued `UnitQuaternion` objects.  Fixed for `RPY()`, `Eul()`, `UnitQuaternion()` constructor, `log()` and `exp()`.  Also had to fix edge case for `log()` and `exp()` for the unit quaternions case.

Tests and doco updated.

The result is a lot of changes, but generally improved functionality for the `UnitQuaternion` class.

### CODE REVIEW

File: spatialmath/pose3d.py [hunk 1]
```diff
@@ -931,6 +931,29 @@ def angdist(self, other: SO3, metric: int = 6) -> Union[float, ndarray]:
...
+
+        :references:
+            - `**Hartley, Trumpf** - "Rotation Averaging" - IJCV 2011 <https://users.cecs.anu.edu.au/~hartley/Papers/PDF/Hartley-Trumpf:Rotation-averaging:IJCV.pdf>`_
+            - `Karcher mean <https://en.wikipedia.org/wiki/Karcher_mean`_
+        """
+
+        eta = tol * np.finfo(float).eps
+        R_mean = self[0]  # initial guess
+        while True:
+            r = np.dstack((self.inv() * self).log()).mean(axis=2)
```
Comment (myeatman-bdai) [inline spatialmath/pose3d.py:952]: It's unclear to me what section of the reference paper outlines this algorithm (admittedly, I did not thoroughly read it) . Would you mind pointing it out to me?

File: tests/test_pose3d.py [hunk 2]
```diff
@@ -717,6 +717,29 @@ def test_functions_lie(self):
...
+    def test_mean(self):
+        rpy = np.ones((100, 1)) @ np.c_[0.1, 0.2, 0.3]
+        R = SO3.RPY(rpy)
+        self.assertEqual(len(R), 100)
+        m = R.mean()
+        self.assertIsInstance(m, SO3)
+        array_compare(m, R[0])
+
+        # now add noise
+        rpy += 0.1 * np.random.rand(100, 3)
```
Comment (myeatman-bdai) [inline tests/test_pose3d.py:739]: This didn't end up changing the rotations, when I add a breakpoint in the mean function for SO3 and inspect the values, there are the same post and pre noise.  I think there's an unexpected copy happening.

File: spatialmath/quaternion.py [hunk 3]
```diff
@@ -1569,6 +1601,29 @@ def dotb(self, omega: ArrayLike3) -> R4:
...
+        :references:
+            - `**Hartley, Trumpf** - "Rotation Averaging" - IJCV 2011 <https://users.cecs.anu.edu.au/~hartley/Papers/PDF/Hartley-Trumpf:Rotation-averaging:IJCV.pdf>`_
+            - `Karcher mean <https://en.wikipedia.org/wiki/Karcher_mean`_
+        """
+
+        eta = tol * np.finfo(float).eps
+        q_mean = self[0]  # initial guess
+        while True:
+            r = (self.inv() * self).log().vec.mean(axis=0)
+            if np.linalg.norm(r) < eta:
```
Comment (myeatman-bdai) [inline spatialmath/quaternion.py:1623]: In the tests, it looks like this case always returns True on the first iteration, can we add a test to exercise the iterative nature of this algorithm?

**Review by myeatman-bdai — CHANGES_REQUESTED** [review]
See inline comments.

File: spatialmath/pose3d.py [hunk 4]
```diff
@@ -931,6 +931,29 @@ def angdist(self, other: SO3, metric: int = 6) -> Union[float, ndarray]:
...
+        :param tol: iteration tolerance in units of eps, defaults to 20
+        :type tol: float, optional
+        :return: the mean rotation
+        :rtype: :class:`SO3` instance.
+
+        Computes the Karcher mean of the set of rotations within the SO(3) instance.
+
+        :references:
+            - `**Hartley, Trumpf** - "Rotation Averaging" - IJCV 2011 <https://users.cecs.anu.edu.au/~hartley/Papers/PDF/Hartley-Trumpf:Rotation-averaging:IJCV.pdf>`_
+            - `Karcher mean <https://en.wikipedia.org/wiki/Karcher_mean`_
```
[the remark this answers is not in the document]
Comment (petercorke) [reply spatialmath/pose3d.py:946]: It should, fixed.

File: spatialmath/pose3d.py (hunk 1 above)
Comment (petercorke) [reply spatialmath/pose3d.py:952]: I've clarified that.

File: tests/test_quaternion.py [hunk 5]
```diff
@@ -1018,6 +1087,19 @@ def test_miscellany(self):
...
+    def test_mean(self):
+        rpy = np.ones((100, 1)) @ np.c_[0.1, 0.2, 0.3]
+        q = UnitQuaternion.RPY(rpy)
+        self.assertEqual(len(q), 100)
+        m = q.mean()
+        self.assertIsInstance(m, UnitQuaternion)
+        nt.assert_array_almost_equal(m.vec, q[0].vec)
+
+        # now add noise
+        rpy += 0.1 * np.random.rand(100, 3)
```
[the remark this answers is not in the document]
Comment (petercorke) [reply tests/test_quaternion.py:1099]: That's a good point.  How would you feel with a seeded random random, so the sequence is guaranteed to always be the same?  Easy enough to do with Numpy.

[the remark this answers is not in the document]
Comment (myeatman-bdai) [reply tests/test_quaternion.py:1099]: sounds good!

**Comment by petercorke** [comment]
Ok, fixed some bugs, tests pass. There is a randomised test but it is reproducible, using a seeded random number generator.

I've commented out, for now, the quaternion mean.  Naively applying the SO(3) algorithms for quaternions didn't work, and I need to do more reading.

**Review by myeatman-bdai — APPROVED** [review]
👍
